File: lib/cgpt/lib/grid.cc
/*
    GPT - Grid Python Toolkit
    Copyright (C) 2020  Christoph Lehner (christoph.lehner@ur.de, https://github.com/lehner/gpt)

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/
#include "lib.h"

std::map<std::string,_grid_cache_entry_> cgpt_grid_cache;
std::map<GridBase*,std::string> cgpt_grid_cache_tag;

EXPORT(create_grid,{
    
    PyObject* _fdimensions, * _precision, * _cb_mask, * _simd_mask, * _mpi;
    void* _parent_grid;
    if (!PyArg_ParseTuple(args, "OOOOOl", &_fdimensions, &_precision,&_cb_mask,&_simd_mask,&_mpi,&_parent_grid)) {
      return NULL;
    }
    
    Coordinate fdimensions, mpi, cb_mask, simd_mask;
    GridBase* parent_grid = (GridBase*)_parent_grid;
    std::string precision;
    
    cgpt_convert(_fdimensions,fdimensions);
    cgpt_convert(_mpi,mpi);
    cgpt_convert(_precision,precision);
    cgpt_convert(_cb_mask,cb_mask);
    cgpt_convert(_simd_mask,simd_mask);
    
    int Nsimd;
    if (precision == "single") {
      Nsimd = vComplexF::Nsimd();
    } else if (precision == "double") {
      Nsimd = vComplexD::Nsimd();
    } else {
      ERR("Unknown precision");
    }

    long nd = cb_mask.size();
    ASSERT(nd == fdimensions.size());
    ASSERT(nd == simd_mask.size());
    ASSERT(nd == mpi.size());
    Coordinate simd(nd,1);

    int nn=Nsimd;
    int i=nd-1;
    while (nn > 1) {
      if (simd_mask[i]) {
	simd[i]*=2;
	nn/=2;
      }
      i--;
      if (i<0)
	i=nd-1;
    }

    GridBase* grid;

    // graceful exit if grid not compatible with mpi/simd
    for (long d=0;d<nd;d++) {
      int cb_factor = cb_mask[d] ? 2 : 1;

      if (fdimensions[d] % (simd[d] * mpi[d] * cb_factor) != 0) {
	ERR("Dimension %d is not consistent:\n"
	    " fdimension = %d\n"
	    " simd = %d\n"
	    " mpi = %d\n"
	    " cb = %d\n",d,fdimensions[d],simd[d],mpi[d],cb_factor);
      }
    }

    grid = cgpt_create_grid(fdimensions,simd,cb_mask,mpi,parent_grid);
    return PyLong_FromVoidPtr(grid);
    
  });



EXPORT(delete_grid,{
    
    void* p;
    if (!PyArg_ParseTuple(args, "l", &p)) {
      return NULL;
    }
    
    cgpt_delete_grid((GridBase*)p);
    return PyLong_FromLong(0);
    
  });



EXPORT(grid_barrier,{
    
    void* p;
    if (!PyArg_ParseTuple(args, "l", &p)) {
      return NULL;
    }
    
    ((GridBase*)p)->Barrier();
    return PyLong_FromLong(0);
    
  });
  


EXPORT(grid_globalsum,{
    
    void* p;
    PyObject* o;
    if (!PyArg_ParseTuple(args, "lO", &p,&o)) {
      return NULL;
    }
    
    GridBase* grid = (GridBase*)p;
    if (PyComplex_Check(o)) {
      ComplexD c;
      cgpt_convert(o,c);
      grid->GlobalSum(c);
      return PyComplex_FromDoubles(c.real(),c.imag());
    } else if (PyFloat_Check(o)) {
      RealD c;
      cgpt_convert(o,c);
      grid->GlobalSum(c);
      return PyFloat_FromDouble(c);
    } else if (PyLong_Check(o)) {
      uint64_t c;
      cgpt_convert(o,c);
      grid->GlobalSum(c);
      return PyLong_FromLong(c);
    } else if (cgpt_PyArray_Check(o)) {
      PyArrayObject* ao = (PyArrayObject*)o;
      int dt = PyArray_TYPE(ao);
      void* data = PyArray_DATA(ao);
      size_t nbytes = PyArray_NBYTES(ao);
      if (dt == NPY_FLOAT32 || dt == NPY_COMPLEX64) {
	grid->GlobalSumVector((RealF*)data, nbytes / 4);
      } else if (dt == NPY_FLOAT64 || dt == NPY_COMPLEX128) {
	grid->GlobalSumVector((RealD*)data, nbytes / 8);
      } else if (dt == NPY_UINT64) {
	grid->GlobalSumVector((uint64_t*)data, nbytes / 8);
      } else {
	ERR("Unsupported numy data type (single, double, csingle, cdouble, uint64 currently allowed)");
      }
      Py_XINCREF(o);
      return o;
    } else {
      ERR("Unsupported object");
    }
  });

EXPORT(grid_get_processor,{
    
    void* p;
    if (!PyArg_ParseTuple(args, "l", &p)) {
      return NULL;
    }
    
    GridBase* grid = (GridBase*)p;
    long srank,sranks;
    cgpt_grid_get_info(grid,srank,sranks);
    long rank = grid->_processor;
    long ranks = grid->_Nprocessors;
    PyObject* coor = cgpt_convert(grid->_processor_coor);
    PyObject* ldims = cgpt_convert(grid->_ldimensions);
    PyObject* gdims = cgpt_convert(grid->_gdimensions);
    return Py_BuildValue("(l,l,N,N,N,l,l)",rank,ranks,coor,gdims,ldims,srank,sranks);
    
  });
 
EXPORT(grid_broadcast,{

    long root;
    PyObject* _data;
    void* p;
    if (!PyArg_ParseTuple(args, "llO", &p,&root,&_data)) {
      return NULL;
    }

    GridBase* grid = (GridBase*)p;

    ASSERT(cgpt_PyArray_Check(_data));
    PyArrayObject* data = (PyArrayObject*)_data;
    
    char* data_p = (char*)PyArray_DATA(data);
    
    long sz = (long)PyArray_NBYTES(data);

    ASSERT(sz < INT_MAX);

    grid->Broadcast((int)root, data_p, (int)sz);
    Py_INCREF(Py_None);
    return Py_None;

  });

EXPORT(grid_exchange,{

    long send_to, recv_from;
    PyObject* _send_data, * _recv_data;
    void* p;
    if (!PyArg_ParseTuple(args, "lllOO", &p,&send_to,&recv_from,&_send_data,&_recv_data)) {
      return NULL;
    }

    GridBase* grid = (GridBase*)p;

    ASSERT(cgpt_PyArray_Check(_send_data) && cgpt_PyArray_Check(_recv_data));
    PyArrayObject* send_data = (PyArrayObject*)_send_data;
    PyArrayObject* recv_data = (PyArrayObject*)_recv_data;
    
    char* send_p = (char*)PyArray_DATA(send_data);
    char* recv_p = (char*)PyArray_DATA(recv_data);
    
    long sz_send = (long)PyArray_NBYTES(send_data);
    long sz_recv = (long)PyArray_NBYTES(recv_data);

    ASSERT(sz_send == sz_recv);
    ASSERT(sz_send < INT_MAX);

    grid->SendToRecvFrom(send_p, (int)send_to, recv_p, (int)recv_from, sz_send);

    Py_INCREF(Py_None);
    return Py_None;

  });

File: lib/gpt/core/grid.py
#
#    GPT - Grid Python Toolkit
#    Copyright (C) 2020  Christoph Lehner (christoph.lehner@ur.de, https://github.com/lehner/gpt)
#
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License along
#    with this program; if not, write to the Free Software Foundation, Inc.,
#    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
#
import cgpt, gpt, sys
import numpy as np


class full:
    def __init__(self, nd):
        self.n = 1
        self.cb_mask = [0] * nd
        self.simd_mask = [1] * nd
        self.__name__ = "full"


class redblack:
    def __init__(self, nd):
        self.n = 2
        rbd = min([nd, 4])
        self.cb_mask = [0] * (nd - rbd) + [1] * rbd
        self.simd_mask = [1] * nd
        self.__name__ = "redblack"


class general:
    def __init__(self, n, cb_mask, simd_mask):
        self.n = n
        self.cb_mask = cb_mask
        self.simd_mask = simd_mask
        self.__name__ = "general_%d_%s_%s" % (n, cb_mask, simd_mask)


def str_to_checkerboarding(s, nd):
    if s == "full":
        return full(nd)
    elif s == "redblack":
        return redblack(nd)
    else:
        a = s.split("_")
        if a[0] == "general":
            n = int(a[1])
            cb_mask = [int(x) for x in a[2].strip("[]").split(",")]
            simd_mask = [int(x) for x in a[3].strip("[]").split(",")]
            return general(n, cb_mask, simd_mask)
        assert 0


def grid_from_description(description):
    p = description.split(";")
    fdimensions = [int(x) for x in p[0].strip("[]").split(",")]
    precision = gpt.str_to_precision(p[1])
    cb = str_to_checkerboarding(p[2], len(fdimensions))
    obj = None
    return grid(fdimensions, precision, cb, obj)


def grid_get_mpi_default(fdimensions, cb):
    nd = len(fdimensions)
    tag = "--mpi"

    # first try to find mpi layout for dimension nd
    mpi = gpt.default.get_ivec(tag, None, nd)
    if mpi is None and nd > 4:

        # if not found but dimension is larger than four, we try to extend the four-dimensional grid
        mpi = gpt.default.get_ivec(tag, None, 4)
        if mpi is not None:
            mpi = [1] * (nd - 4) + mpi

    if mpi is None:
        # try trivial layout
        mpi = [1] * nd

    assert nd == len(mpi)
    return mpi


class grid:
    def __init__(
        self, fdimensions, precision, cb=None, obj=None, mpi=None, parent=None
    ):

        self.fdimensions = fdimensions
        self.fsites = np.prod(self.fdimensions)
        self.precision = precision
        self.nd = len(self.fdimensions)

        if cb is None:
            cb = full

        if isinstance(cb, type):
            cb = cb(self.nd)

        self.cb = cb
        if mpi is None:
            # if we live on a split grid, cannot use default mpi layout
            assert parent is None
            self.mpi = grid_get_mpi_default(self.fdimensions, self.cb)
        else:
            self.mpi = mpi

        self.parent = parent
        if parent is None:
            parent_obj = 0
        else:
            parent_obj = parent.obj

        if obj is None:
            self.obj = cgpt.create_grid(
                fdimensions, precision, cb.cb_mask, cb.simd_mask, self.mpi, parent_obj
            )
        else:
            self.obj = obj

        # processor is mpi rank, may not be lexicographical (cartesian) rank
        (
            self.processor,
            self.Nprocessors,
            self.processor_coor,
            self.gdimensions,
            self.ldimensions,
            self.srank,
            self.sranks,
        ) = cgpt.grid_get_processor(self.obj)
        self.gsites = np.prod(self.gdimensions)

    def describe(
        self,
    ):  # creates a string without spaces that can be used to construct it again, this should only describe the grid geometry not the mpi
        s = (
            str(self.fdimensions)
            + ";"
            + self.precision.__name__
            + ";"
            + self.cb.__name__
        ).replace(" ", "")
        if self.parent is not None:
            s += ";" + self.parent.describe()
        return s

    def converted(self, dst_precision):
        if dst_precision == self.precision:
            return self
        if self.parent is None:
            parent = None
        else:
            parent = self.parent.converted(dst_precision)
        return grid(
            self.fdimensions,
            dst_precision,
            cb=self.cb,
            obj=None,
            mpi=self.mpi,
            parent=parent,
        )

    def checkerboarded(self, cb):
        if cb == self.cb:
            return self
        if self.parent is None:
            parent = None
        else:
            parent = self.parent.checkerboarded(cb)
        return grid(
            self.fdimensions,
            self.precision,
            cb=cb,
            obj=None,
            mpi=self.mpi,
            parent=parent,
        )

    def split(self, mpi_split, fdimensions):
        return grid(fdimensions, self.precision, self.cb, None, mpi_split, self)

    def inserted_dimension(self, dimension, extent, cb_mask=None, simd_mask=1):
        if cb_mask is None and self.cb.n == 1:
            cb_mask = 0
        assert cb_mask is not None
        cb = general(
            self.cb.n,
            self.cb.cb_mask[0:dimension] + [cb_mask] + self.cb.cb_mask[dimension:],
            self.cb.simd_mask[0:dimension]
            + [simd_mask]
            + self.cb.simd_mask[dimension:],
        )

        if self.parent is None:
            parent = None
            mpi = None
        else:
            parent = self.parent.inserted_dimension(dimension, extent)
            mpi = self.mpi[0:dimension] + [1] + self.mpi[dimension:]

        return grid(
            self.fdimensions[0:dimension] + [extent] + self.fdimensions[dimension:],
            self.precision,
            cb=cb,
            obj=None,
            mpi=mpi,
            parent=parent,
        )

    def removed_dimension(self, dimension):
        assert 0 <= dimension and dimension < self.nd
        cb = general(
            self.cb.n,
            self.cb.cb_mask[0:dimension] + self.cb.cb_mask[dimension + 1 :],
            self.cb.simd_mask[0:dimension] + self.cb.simd_mask[dimension + 1 :],
        )

        if self.parent is None:
            parent = None
        else:
            parent = self.parent.removed_dimension(dimension)

        return grid(
            self.fdimensions[0:dimension] + self.fdimensions[dimension + 1 :],
            self.precision,
            cb=cb,
            obj=None,
            mpi=None,
            parent=parent,
        )

    def cartesian_rank(self):
        rank = 0
        for i in reversed(range(self.nd)):
            rank = rank * self.mpi[i] + self.processor_coor[i]
        return rank

    def __str__(self):
        s = f"fdimensions = {self.fdimensions}; mpi = {self.mpi}; precision = {self.precision.__name__}; checkerboard = {self.cb.__name__}"
        if self.parent is not None:
            s += " split from " + self.parent.__str__()
        return s

    def __del__(self):
        cgpt.delete_grid(self.obj)

    def barrier(self):
        cgpt.grid_barrier(self.obj)

    def globalsum(self, x):
        if type(x) == gpt.tensor:
            cgpt.grid_globalsum(self.obj, x.array)
            return x
        else:
            return cgpt.grid_globalsum(self.obj, x)
